backend/server.py

```
@@ -46,52 +49,120 @@
+                    signal.signal(signal.SIGALRM, old_handler)
+                return result
+            else:
+                # Windows fallback: use ThreadPoolExecutor for timeout
+                with ThreadPoolExecutor(max_workers=1) as executor:
+                    future = executor.submit(f, *args, **kwargs)
+                    try:
+                        return future.result(timeout=timeout_seconds)
+                    except FuturesTimeoutError:
+                        raise MLTimeoutError(
+                            f"Operation timed out after {timeout_seconds} seconds"
+                        )
```

Ruff (0.15.7)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)